### DISCUSSION TITLE
Get previous lexeme in external scanner

### DISCUSSION THREAD

**Discussion by cfroystad** [tree-sitter/tree-sitter#1339] [Q&A] [answered]
I'm [struggling with parsing PHP strings](https://github.com/tree-sitter/tree-sitter-php/pull/72#issuecomment-898859797), and I currently see no other solution than to make the external scanner behave differently depending on which lexeme precedes the scanning of the PHP string.

Thus, my question is: Is it possible to get the preceding lexeme in the external scanner? If so, how would I go about doing this?

**Comment by ahelwer** [comment, marked as answer]
Yes, I do this for my external scanner. What you have to do is move the preceding token into the external scanner, then maintain a stateful flag like `has_seen_token`. You can see [my external scanner](https://github.com/tlaplus-community/tree-sitter-tlaplus/blob/6c6a935634c696ab728d315a273a387e1e7c2c52/src/scanner.cc#L1481) where I do this for the `PROOF_KEYWORD` token. You can see the corresponding use of this token on the grammar side [here](https://github.com/tlaplus-community/tree-sitter-tlaplus/blob/6c6a935634c696ab728d315a273a387e1e7c2c52/grammar.js#L997).

Alternatively, if possible you can handle this at the grammar level with rules analogous to the following:
```js
rule: $ => choice(
  seq($.lexeme, $.external_token_1),
  $.external_token_2
),
```
then branch based on the valid symbols in your external scanner. Of course, `external_token_2` would have to be able to detect `lexeme` and then return false.

A third option is similar to the first, but instead of emitting the preceding token from the external scanner you emit an invisible marker token then handle the lexeme in your grammar. In this way your external scanner gets notified of the presence of the lexeme (presumably saving this information in a `has_seen_token` flag) but you can handle the lexeme in your grammar in case you want to break it up into further tokens; something like:
```js
rule: $ => seq(optional(seq($._external_invisible_marker, $.lexeme)), $.external_token),
```
Note that tree-sitter allows external tokens of size zero, which admits the danger of infinite loops but shows its utility here.
I do something like this with my invisible external [`_begin_proof_step`](https://github.com/tlaplus-community/tree-sitter-tlaplus/blob/6c6a935634c696ab728d315a273a387e1e7c2c52/grammar.js#L1014) token, which always precedes the [`proof_step_id`](https://github.com/tlaplus-community/tree-sitter-tlaplus/blob/6c6a935634c696ab728d315a273a387e1e7c2c52/grammar.js#L1093) lexeme which I break down into a series of addressable tokens. The only drawback of this approach is you scan the same lexeme multiple times, which might be a performance issue.

**Reply by cfroystad** [reply]
Thank you very much for sharing your knowledge! This gives me a few very good options which I'd never come to think about on my own 🙂

**Reply by ahlinc** [reply]
@cfroystad It would be worth to add in the #1140 discussion a link to here and mark it resolved, I suppose.
